src/util/ThreadPriority.cpp

```diff
@@ -0,0 +1,134 @@
+/// The maximum priority level for the SCHED_RR policy
+const int max_rr_priority = sched_get_priority_max(SCHED_RR);
+/// The maximum priority level for the SCHED_FIFO policy
+const int max_fifo_priority = sched_get_priority_max(SCHED_FIFO);
+
+void set_priority(const NUClear::PriorityLevel& priority) {
+
+    sched_param param{};
+    switch (priority) {
+        case NUClear::PriorityLevel::IDLE: pthread_setschedparam(pthread_self(), SCHED_IDLE, &param); break;
+        case NUClear::PriorityLevel::LOWEST:
+        case NUClear::PriorityLevel::LOW: pthread_setschedparam(pthread_self(), SCHED_OTHER, &param); break;
+        default:  // Default to normal if someone broke the enum
+        case NUClear::PriorityLevel::NORMAL:
+            param.sched_priority = min_rr_priority;
+            pthread_setschedparam(pthread_self(), SCHED_RR, &param);  // Min realtime beats all non-realtime
+            break;
+        case NUClear::PriorityLevel::HIGH:
+            param.sched_priority = (min_rr_priority + max_rr_priority + 1) / 2;  // Halfway between min and max
+            pthread_setschedparam(pthread_self(), SCHED_RR, &param);
+            break;
+        case NUClear::PriorityLevel::HIGHEST:
+            param.sched_priority = max_rr_priority - 1;  // One less than max so it can be preempted by realtime
+            pthread_setschedparam(pthread_self(), SCHED_RR, &param);
+            break;
+        case NUClear::PriorityLevel::REALTIME:
+            param.sched_priority = max_fifo_priority;  // Max priority for SCHED_FIFO
+            pthread_setschedparam(pthread_self(), SCHED_FIFO, &param);
+            break;
+    }
+}
+
+}  // namespace
+
+#elif defined(__APPLE__)
+
+    #include <pthread.h>
+
+namespace {
+
+void set_priority(const NUClear::PriorityLevel& priority) {
+    switch (priority) {
+        case NUClear::PriorityLevel::IDLE: pthread_set_qos_class_self_np(QOS_CLASS_BACKGROUND, 0); break;
+        case NUClear::PriorityLevel::LOWEST: pthread_set_qos_class_self_np(QOS_CLASS_UTILITY, 1); break;
+        case NUClear::PriorityLevel::LOW: pthread_set_qos_class_self_np(QOS_CLASS_UTILITY, 0); break;
+        default:  // Default to normal if someone broke the enum
+        case NUClear::PriorityLevel::NORMAL: pthread_set_qos_class_self_np(QOS_CLASS_DEFAULT, 0); break;
+        case NUClear::PriorityLevel::HIGH: pthread_set_qos_class_self_np(QOS_CLASS_USER_INITIATED, 1); break;
+        case NUClear::PriorityLevel::HIGHEST: pthread_set_qos_class_self_np(QOS_CLASS_USER_INITIATED, 0); break;
+        case NUClear::PriorityLevel::REALTIME: pthread_set_qos_class_self_np(QOS_CLASS_USER_INTERACTIVE, 0); break;
+    }
+}
+
+}  // namespace
+
+#else
+    #error "Unsupported platform"
+#endif
+
+namespace NUClear {
+namespace util {
+
+    ThreadPriority::ThreadPriority(const NUClear::PriorityLevel& priority) : previous_priority(current_priority) {
+        if (previous_priority != current_priority) {
```

Comment:
```suggestion
        if (priority != current_priority) {
```

Reply:
Also, you should unit test this class to make sure that the lock works the way you expect